### DISCUSSION TITLE
Qos Settings

### DISCUSSION THREAD

**Discussion by MuhammedErkam** [OpenDDS/OpenDDS#3870] [Q&A] [answered]
I am currently working on the StockQuoter Example.

In this example I have changed the Qos settings of DataWriter and DataReader. 

But I couldn't understand some settings.

**First** what is the importance of DURABILITY_SERVICE Qos settings because I changed it but didn't see any difference.

**Second** when I use RELIABILITY MODE, If max_sample is greater than multiply of the max_instance and max_sample_per_instance then system works, but If max_sample is less than that multiplication, BUT if I give max_samples less than multiplication the system will only generate samples as many as max_samples.

For example, Let's say I am using RELIABILITY MODE and history.depth is 5. 

And I am publishing my samples in below for loop:

```
for ( int i = 0; i < 40; ++i ) {
    StockQuoter::Quote spy_quote;
    spy_quote.exchange = STOCK_EXCHANGE_NAME;
    spy_quote.ticker = CORBA::string_dup("SPY");
    spy_quote.full_name = CORBA::string_dup("S&P Depository Receipts");
    spy_quote.value = 0.0 + 1.0*i;
    spy_quote.timestamp = get_timestamp();

    cout << "Publishing SPY Quote: " << spy_quote.value << endl;
    ret = quote_dw->write(spy_quote, spy_handle);
    if (ret != DDS::RETCODE_OK) {
        ACE_ERROR ((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: SPY write returned %d.\n"), ret));
    }

    StockQuoter::Quote mdy_quote;
    mdy_quote.exchange = STOCK_EXCHANGE_NAME;
    mdy_quote.ticker = CORBA::string_dup("MDY");
    mdy_quote.full_name = CORBA::string_dup("S&P Midcap Depository Receipts");
    mdy_quote.value = 100.0 + 1.0*i;
    mdy_quote.timestamp = get_timestamp();

    cout << "Publishing MDY Quote: " << mdy_quote.value <<endl;
    ret = quote_dw->write(mdy_quote, mdy_handle);
    if (ret != DDS::RETCODE_OK) {
        ACE_ERROR ((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: MDY write returned %d.\n"), ret));
    }

    ACE_OS::sleep( quarterSecond );
}
```

If I do below settings for RESOURCE_LIMITS THAT'S OK.

```
dataWriterQos.resource_limits.max_instances = 2;
dataWriterQos.resource_limits.max_samples_per_instance = 5;             
dataWriterQos.resource_limits.max_samples = 10;   
```

BUT if I give max_samples less than 10 the system will only generate samples as many as max_samples.
For example, If I say max_samples = 9, when 9 samples are produced, system is not working.

**Third** I do not clearly undestand importance of Topic QoS. What does it set? Do you have any suggestions for setting it?

**Fourth** Do you have any suggestions for setting RESOURCE LIMITS as there are no suggestions in the Developer Guide?

**Comment by mitza-oci** [comment, marked as answer]
1. See section 3.2.6 of the OpenDDS Developer's Guide
2. I agree that max_samples limits that overall number of samples that can be stored in the Writer's cache and this is by design.  If you're seeing a behavior that doesn't match the docs/spec it's best to add a test to the repository.
3. I suggest ignoring Topic QoS.  It's not necessary for setting the policies of the system.
4. The resource limits are useful for putting upper bounds on memory consumption on systems where this is important
